### ISSUE TITLE
Support custom types in conversion

### ISSUE THREAD

**Issue by ianhlavats** [drkameleon/iswift#129]
When attempting to convert this code, iSwift fails:

```
@implementation someClass
    - (void)someFunc {
        MyClass * foo = nil;
    }
@end
```
See screenshot:

![screen shot 2015-11-18 at 9 32 58 pm](https://cloud.githubusercontent.com/assets/3005952/11253461/fdf7df92-8e3b-11e5-9287-8e58051707c6.png)

Expected result:

```
class someClass {
    func someFunc() {
        var foo: MyClass? = nil
    }
}
```

**Comment by drkameleon** [comment]
Re-visiting the other issue, let me put it this way:

- What is `MyClass` ? Is it a random word? Is it a type?

Now, given that iSwift is actually a - let's say mini-version, but still one - compiler, would that compile in Xcode?

No, because it wouldn't know what "MyClass" is.

A few ways to make it "know" about it (that it is a *type identifier*) : 

- Add a `@class MyClass;` statements on top of your code.
- Add a header import, like `#import "MyClass.h"` (it won't "import" the header. It just guesses it's a type)
- Add all above statements in a "header" file which you then pre-import (have a look under preferences - whatevern you put there, if enabled, will always be as if it was pasted at the top of your code.)

And there's no universal way to do that. If it's a known type (e.g. `NSString`) then it should be automatically recognized as such (there are currently some thousands of types already known, so in case you spot any type not being recognized, just let me know). But for user types, there's simply no way. You have to let it know somehow that it's a user type. ;-)

**Comment by ianhlavats** [comment]
I suppose the way iSwift works is tightly coupled to the Swift compiler and therefore has this limitation. Can you not pre-process the code before feeding it to the compiler? For example generate the class definition on the fly or simply comment out that code and then use some simple string replacement logic to convert it from Objective-C to Swift? Clearly it's possible to convert user defined classes from Objective-C to Swift... objectivec2swift.com has no issues converting them. If this limitation must exist with your tool don't you think it will significantly limit it's usefulness for converting real-world code?

**Comment by drkameleon** [comment]
I understand what you mean in terms of flexibility (and btw, you gave me an idea about how it *might* - just might - work, but I'll see it as it goes).

Re: iSwift inner workings: no, it's not coupled to the Swift compiler or the Objective-C/LLVM or any compiler for that matter. It's a "compiler" in its own right. And as with all compilers (take a source code, parse it, convert it into a tree, generate some code - binary or not - etc), it conforms to some rules and grammar, hence it throws errors whenever something is not according to the correct grammar specs. So, in a few words, It's not anywhere as simplistic as a text-replacement. It *could* have been (in theory), but this way the results would have been REALLY poor, with no output code optimization options - not to mention, that I doubt whether that'd be feasible at all. 

Re: the other tool you mention, although they have a parser/compiler too - they still allow errors. That's the main difference in terms of parsing. Now... is that bad? Well.. It may be good (as in the case we're talking about). But it may also not be the case at other points. (More if you ask me. Except for the issue with unknown user-defined types, a more stricter... error tracking yields far better results, I believe...  - But I guess that's debatable. So, all ideas are welcome! :))

**Comment by drkameleon** [comment]
Re:

> don't you think it will significantly limit it's usefulness for converting real-world code

Well, if what can be compiled in Xcode is real enough (for me, at least, that's as real as it can get - and it has been the basis of all my "experiments"), I really cannot understand the limitation. I can acknowledge the fact that iSwift doesn't support header imports (i mean... *actually* importing the header) YET, but it will most definitely do very soon. My main goal is to have every valid Xcode project automatically translated into Swift. As simple as that. ;-)

**Comment by ianhlavats** [comment]
Thanks for your reply. I think this touches an important concept, is iSwift a code conversion tool or code validation tool? The Swift compiler built into Xcode already does the job of catching errors at compile time. What I am looking for I guess in a tool like iSwift is strictly code conversion from one language to another. 

To use an analogy, just like Google Translate sometimes makes mistakes when translating from English to another language, that's fine, because I can correct that. But if Google Translate refused to translate my English into another language because I used a word it did know, let's just say it would not be the best translation tool out there. ;) It happens to work well enough to "convert" my text into another language, and if I am proficient enough in that language I can take it from there.

I would like to see the same thing from iSwift. At least give me the option to disable the code validation... it would be much faster anyway to treat the text as a string replacement exercise. Some amount of optimization can still happen, like using let instead of var etc. I think the problem you are trying to solve here is really about converting from one language's syntax to another. 

I get that you are building an AST internally and need some conformity to a language standard in order to parse and process the code, but currently iSwift needs work to be able to solve the real-world programming problems of converting blocks of code and entire files from Objective-C to Swift. I am trying to be constructive and would be happy to keep testing the product (I did pay for a license after all).
